Framework/DataObjects/inc/MantidDataObjects/WorkspaceCreation.h

```diff
@@ -0,0 +1,207 @@
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+namespace detail {
+MANTID_DATAOBJECTS_DLL HistogramData::Histogram
+stripData(HistogramData::Histogram histogram);
+
+template <class T> std::unique_ptr<T> createHelper() {
+  return Kernel::make_unique<T>();
+}
+
+template <class T> std::unique_ptr<T> createConcreteHelper() {
+  return Kernel::make_unique<T>();
+}
+
+template <>
+MANTID_DATAOBJECTS_DLL std::unique_ptr<API::HistoWorkspace> createHelper();
+
+// Dummy specialization, should never be called, must exist for compilation.
+template <>
+MANTID_DATAOBJECTS_DLL std::unique_ptr<EventWorkspace> createHelper();
+// Dummy specialization, should never be called, must exist for compilation.
+template <>
+MANTID_DATAOBJECTS_DLL std::unique_ptr<API::MatrixWorkspace> createHelper();
+
+// Dummy specialization, should never be called, must exist for compilation.
+template <>
+MANTID_DATAOBJECTS_DLL std::unique_ptr<API::MatrixWorkspace>
+createConcreteHelper();
+// Dummy specialization, should never be called, must exist for compilation.
+template <>
+MANTID_DATAOBJECTS_DLL std::unique_ptr<API::HistoWorkspace>
+createConcreteHelper();
+}
+
+template <class T, class P, class IndexArg,
+          class = typename std::enable_if<
+              std::is_base_of<API::MatrixWorkspace, P>::value>::type>
+boost::shared_ptr<T> create(const P &parent, const IndexArg &indexArg,
+                            const HistogramData::Histogram &histogram) {
+  // Figure out (dynamic) target type:
+  // - Type is same as parent if T is base of parent
+  // - If T is not base of parent, conversion may occur. Currently only
+  //   supported for EventWorkspace
+  boost::shared_ptr<T> ws;
+  if (std::is_base_of<API::HistoWorkspace, T>::value &&
+      parent.id() == "EventWorkspace") {
+    // Drop events, create Workspace2D or T whichever is more derived.
+    ws = detail::createHelper<T>();
+  } else {
+    try {
+      // If parent is more derived than T: create type(parent)
+      ws = dynamic_cast<const T &>(parent).cloneEmpty();
```

Comment:
Let me give an example:

```cpp
const MatrixWorkspace &parent = ... // say Workspace2D
auto child = create<MatrixWorkspace>(parent);
```

The *dynamic* type of parent is more derived than `T` (which is `MatrixWorkspace`), `cloneEmpty()` will create the correct type.

The other case:

```cpp
const MatrixWorkspace &parent = ... // say Workspace2D
auto child = create<SpecialWorkspace2D>(parent);
```

T is more derived than the parent, so the cast will fail, and we enter the other code branch via `throw` + `catch`. This is not particularly nice, but the alternative would be to attempt some string matching based on the workspace ID, which is also ugly, I think?